doc/api/module.md

```diff
@@ -269,80 +164,239 @@ changes:
 
 ```bash
-node \
-  --experimental-loader unpkg \
-  --experimental-loader http-to-https \
-  --experimental-loader cache-buster
+node --import 'data:text/javascript,import { register } from "node:module"; register("http-to-https", import.meta.url);' ./my-app.js
 ```
 
-These are called in the following sequence: `cache-buster` calls
-`http-to-https` which calls `unpkg`.
+### Chaining
+
+It's possible to call `register` more than once:
+
+```mjs
+// entrypoint.mjs
+import { register } from 'node:module';
+
+register(new URL('./first.mjs', import.meta.url));
+register('./second.mjs', import.meta.url);
+await import('./my-app.mjs');
+```
+
+```cjs
+// entrypoint.cjs
+const { register } = require('node:module');
+const { pathToFileURL } = require('node:url');
+
+const parentURL = pathToFileURL(__filename);
+register(new URL('./first.mjs', parentURL));
+register('./second.mjs', parentURL);
+import('./my-app.mjs');
+```
+
+In this example, the registered hooks will form chains. If both `first.mjs` and
+`second.mjs` define a `resolve` hook, both will be called, in the order they
+were registered. The same applies to all the other hooks.
+
+### Communication between main and hooks threads
+
+Module customization hooks run on a dedicated thread, separate from the main
+thread that runs application code. This means that global variables are unique
+to each thread, and message channels must be used to communicate between the
+threads.
+
+The `register` method can be used to pass data to an [`initialize`][] hook on
+the hooks thread. The data passed to the hook may include transferrable objects
+like ports.
+
+```mjs
+import { register } from 'node:module';
+import { MessageChannel } from 'node:worker_threads';
+
+// This example showcases how a message channel can be used to
+// communicate to the hooks, by sending `port2` to the hooks.
+const { port1, port2 } = new MessageChannel();
+
+port1.on('message', (msg) => {
+  console.log(msg);
+});
+
+register('./my-hooks.mjs', {
+  parentURL: import.meta.url,
+  data: { number: 1, port: port2 },
+  transferList: [port2],
+});
+```
+
+```cjs
+const { register } = require('node:module');
+const { pathToFileURL } = require('node:url');
+const { MessageChannel } = require('node:worker_threads');
+
+// This example showcases how a message channel can be used to
+// communicate to the hooks, by sending `port2` to the hooks.
+const { port1, port2 } = new MessageChannel();
+
+port1.on('message', (msg) => {
+  console.log(msg);
+});
+
+register('./my-hooks.mjs', {
+  parentURL: pathToFileURL(__filename),
+  data: { number: 1, port: port2 },
+  transferList: [port2],
+});
+```
 
 ### Hooks
 
+The [`register`][] method can be used to register a module that exports a set of
+hooks. The hooks are functions that are called by Node.js to customize the
+module resolution and loading process. The exported functions must have specific
+names and signatures, and they must be exported as named exports from an ES
+module file.
```

Comment:
Should we start to deprecate using CJS on this loader thread? Currently exporting from a CJS module works just fine.

Reply:
We’ve never documented that hooks files could be CommonJS, so I think there’s nothing to deprecate? I’d rather not document it now because then it becomes a feature we need to continue to support (or deprecate) whereas in the current state we can remove that ability in the future (or more likely, inadvertently break that ability) and we haven’t created a bug.

Either way, I think this is a question beyond the scope of this PR. If we’re going to document this support then we should also add a test for it, which feels like a separate PR from this one.

Reply:
I think it's a different thing to have omissions in the docs (e.g. not mentioning that a CJS loader module works), and to have factually incorrect documentation (they must NOT be exported from an ES module file, it can be a CJS module, or a `data:`-URL module, or even HTTP with the ad-hoc flag). I suggest remove the last part of the sentence.
```suggestion
names and signatures, and they must be exported as named exports.
```